pkg/resources/cloudcontroller/hetzner.go

```diff
@@ -0,0 +1,139 @@
+limitations under the License.
+*/
+
+package cloudcontroller
+
+import (
+	"fmt"
+
+	"k8c.io/kubermatic/v2/pkg/resources"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	"k8c.io/kubermatic/v2/pkg/resources/vpnsidecar"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	HetznerCCMDeploymentName = "hcloud-cloud-controller-manager"
+)
+
+var (
+	hetznerResourceRequirements = corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse("50Mi"),
+			corev1.ResourceCPU:    resource.MustParse("100m"),
+		},
+	}
+)
+
+func hetznerDeploymentCreator(data *resources.TemplateData) reconciling.NamedDeploymentCreatorGetter {
+	return func() (string, reconciling.DeploymentCreator) {
+		return HetznerCCMDeploymentName, func(dep *appsv1.Deployment) (*appsv1.Deployment, error) {
+			dep.Labels = resources.BaseAppLabels(HetznerCCMDeploymentName, nil)
+
+			dep.Spec.Replicas = resources.Int32(1)
+
+			dep.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: resources.BaseAppLabels(HetznerCCMDeploymentName, nil),
+			}
+
+			podLabels, err := data.GetPodTemplateLabels(HetznerCCMDeploymentName, dep.Spec.Template.Spec.Volumes, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			dep.Spec.Template.ObjectMeta = metav1.ObjectMeta{
+				Labels: podLabels,
+			}
+
+			dep.Spec.Template.Spec.DNSPolicy, dep.Spec.Template.Spec.DNSConfig, err =
+				resources.UserClusterDNSPolicyAndConfig(data)
+			if err != nil {
+				return nil, err
+			}
+
+			f := false
+			dep.Spec.Template.Spec.AutomountServiceAccountToken = &f
+
+			openvpnSidecar, err := vpnsidecar.OpenVPNSidecarContainer(data, "openvpn-client")
+			if err != nil {
+				return nil, fmt.Errorf("failed to get openvpn sidecar: %v", err)
+			}
+
+			credentials, err := resources.GetCredentials(data)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get credentials: %v", err)
+			}
+
+			network := data.Cluster().Spec.Cloud.Hetzner.Network
+			if network == "" {
+				network = data.DC().Spec.Hetzner.Network
+			}
+
+			dep.Spec.Template.Spec.Volumes = getVolumes()
+
+			dep.Spec.Template.Spec.Containers = []corev1.Container{
+				*openvpnSidecar,
+				{
+					Name:  HetznerCCMDeploymentName,
+					Image: data.ImageRegistry(resources.RegistryDocker) + "/hetznercloud/hcloud-cloud-controller-manager:v1.8.1",
+					Command: []string{
+						"/bin/hcloud-cloud-controller-manager",
+						"--kubeconfig=/etc/kubernetes/kubeconfig/kubeconfig",
+						"--cloud-provider=hcloud",
+						"--leader-elect=false",
+						"--allow-untagged-cloud",
+						"--allocate-node-cidrs=true",
+						"--cluster-cidr=10.244.0.0/16",
+					},
+					Env: []corev1.EnvVar{
+						{
+							Name: "NODE_NAME",
+							ValueFrom: &corev1.EnvVarSource{
+								FieldRef: &corev1.ObjectFieldSelector{
+									FieldPath: "spec.nodeName",
+								},
+							},
```

Comment:
Is that actually correct? Since we're running CCM in the seed cluster in the user cluster namespace, that's going to equal to a node in the seed cluster (which might not be a Hetzner node at all, it could be GKE, AWS, etc.).

I'm not sure how Hetzner CCM uses that value (it's required), but this might eventually lead to unexpected behavior.

Reply:
This is a fun one. As you mentioned, it's a mandatory env variable, which is then later used to determine the current zone from the machine the CCM is running on. _However_, this is currently not used anywhere in the CCM, and only the unit tests use the GetZone() functions.

So technically for now it doesn't matter what value the env variable has. However this might change and makes me wonder if we need to deploy the CCM into the usercluster instead. I could also imagine us making an upstream PR to allow a new `HCLOUD_ZONE` variable, which would remove the need for dynamically determining the zone based on a hostname.

Reply:
> However, this is currently not used anywhere in the CCM, and only the unit tests use the GetZone() functions.

This seems to be true.

> However this might change and makes me wonder if we need to deploy the CCM into the usercluster instead. I could also imagine us making an upstream PR to allow a new HCLOUD_ZONE variable, which would remove the need for dynamically determining the zone based on a hostname.

For the beginning, can you make an issue in the Hetzner CCM repo? We could discuss this with them, and eventually, make a PR to add a new variable. If it doesn't work for them, we can try to deploy it in the user cluster.

Reply:
https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/159